File: kit-fastapi/app/application/todo/create_todo.py
"""Cas d'usage : créer une tâche. Valide, construit l'entité, persiste."""
from __future__ import annotations

from app.domain.todo import rules
from app.domain.todo.clock import Clock
from app.domain.todo.exceptions import TodoValidationError
from app.domain.todo.id_generator import IdGenerator
from app.domain.todo.repository import TodoRepository
from app.domain.todo.todo import Todo


class CreateTodo:
    def __init__(self, todos: TodoRepository, clock: Clock, ids: IdGenerator) -> None:
        self._todos = todos
        self._clock = clock
        self._ids = ids

    def execute(self, data: dict) -> Todo:
        errors = rules.validate(data)
        if errors:
            raise TodoValidationError(errors)
        now = self._clock.now()
        todo = Todo(
            id=self._ids.uuid(),
            title=rules.normalize_title(str(data["title"])),
            done=False,
            created_at=now,
            updated_at=now,
        )
        self._todos.save(todo)
        return todo

File: kit-php/src/Application/Todo/CreateTodo.php
<?php

declare(strict_types=1);

namespace TodoApp\Application\Todo;

use TodoApp\Domain\Todo\Clock;
use TodoApp\Domain\Todo\IdGenerator;
use TodoApp\Domain\Todo\Todo;
use TodoApp\Domain\Todo\TodoRepository;
use TodoApp\Domain\Todo\TodoRules;
use TodoApp\Domain\Todo\TodoValidationException;

/** Cas d'usage : créer une tâche. Valide, construit l'entité, persiste. */
final class CreateTodo
{
    public function __construct(
        private readonly TodoRepository $todos,
        private readonly Clock $clock,
        private readonly IdGenerator $ids,
    ) {
    }

    /**
     * @param array<string,mixed> $input
     * @throws TodoValidationException
     */
    public function execute(array $input): Todo
    {
        $errors = TodoRules::validate($input);
        if ($errors !== []) {
            throw new TodoValidationException($errors);
        }

        $now = $this->clock->now();
        $todo = new Todo(
            $this->ids->uuid(),
            TodoRules::normalizeTitle((string) $input['title']),
            false,
            $now,
            $now,
        );

        $this->todos->save($todo);

        return $todo;
    }
}
